### PR TITLE
feat(cff-10): Add "Select all dependabot pull requests"

### PR DESCRIPTION
**Pull request by jisung-lee7**
## Issue Link

[issue-10](https://github.com/kafelix496/coding-for-fun/issues/10)

## Notes

-Use linear instead of github project.

### CODE REVIEW

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 1]
```diff
@@ -48,6 +93,25 @@ export default function Page() {
...
+        >
+          <Button
+            variant="ghost"
+            size="icon"
+            icon={
+              <Tooltip tooltip={'NOT DECIDED'}>
+                <SettingsIcon className="w-full h-full" />
+              </Tooltip>
+            }
+          />
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:113]: What is this button's purpose?

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:113]: Hmm? What do you mean? That just makes the SettingIcon work. I was wondering,,, Am I wrong?

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 2]
```diff
@@ -48,6 +114,25 @@ export default function Page() {
...
+            groups: [{ label: '', items: values }],
+            value: '',
+            onValueChange: handleValueChange
+          }}
+        >
+          <Button
+            variant="ghost"
+            size="icon"
+            icon={
+              <Tooltip tooltip={'NOT DECIDED'}>
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:130]: Use translation. Also, tooltip message should be "Settings" and "설정"

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 3]
```diff
@@ -29,6 +38,62 @@ export default function Page() {
...
+    {
+      label: translate('GITHUB.SELECT_ALL_DEPENDABOT_PULL_REQUESTS'),
+      value: ESettingsCode.SELECT_ALL_DEPENDABOT_PULL_REQUESTS,
+      disabled: repos === undefined
+    },
+
+    {
+      label: translate('GITHUB.START_REVIEW_BUTTON'),
+      value: ESettingsCode.START_REVIEW_BUTTON,
+      disabled: flattenCheckedPulls.length <= 0
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:73]: You can just do `flattenCheckedPulls.length === 0`. There is no use-case that length is smaller than 0. Even if there is a use-case, it's already a critical error so we don't have to deal with that.

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 4]
```diff
@@ -29,6 +38,62 @@ export default function Page() {
     openAllRepo()
   }
 
+  const handleSelectAllDependabotClick = () => {
+    if (repos) {
+      repos.map((repo) => {
+        console.log(`c.log ## repo ##`, repo)
+        if (repo.pulls) {
+          repo.pulls.map((pull) => {
+            if (pull.user.login === 'jisung-lee7') {
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:47]: I would use const for this.

```typescript
if (pull.user.login === DEPENDABOT_USER_NAME) {
  // blah blah
}
```

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 5]
```diff
@@ -29,6 +38,62 @@ export default function Page() {
     openAllRepo()
   }
 
+  const handleSelectAllDependabotClick = () => {
+    if (repos) {
+      repos.map((repo) => {
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:43]: I believe you don't know the difference between `map` and `forEach`. Check it out.

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:43]: No,, No believe!! I know the difference between map and forEach!! TT😢

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx (hunk 4 above)
Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:47]: Oh cool

Comment (jiyeol-lee) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:48]: You did not resolved this

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:48]: Ah~ 'const for this', okay. I missed it.

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 6]
```diff
@@ -14,12 +17,19 @@ import {
...
+}
+
 export default function Page() {
   const { repos } = useRepos()
   const flattenCheckedPulls = getFlattenCheckedPulls(repos)
   const { translate } = useDictionary()
   const [isDialogOpen, setIsDialogOpen] = useState<boolean>(false)
-  const { openAllRepo } = useUpdateRepoOrPull()
+  const { openAllRepo, selectAllDependabot } = useUpdateRepoOrPull()
+  const DEPENDABOT_USER_NAME = 'jisung-lee7'
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:32]: put this outside of the `Page` function.

Comment (jiyeol-lee) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:32]: It doesn't have to be inside of the component

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:32]: Hmm,, What do you mean,,? I can't understand.

Comment (jiyeol-lee) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:32]:
```typescript
const DEPENDABOT_USER_NAME = 'jisung-lee7'

export default function Page() {
  // no need to put the const variable inside the component
}
```

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:32]: Ah~ only `const DEPENDABOT_USER_NAME = 'jisung-lee7'`
I understood every const. lol.

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 7]
```diff
@@ -29,25 +40,83 @@ export default function Page() {
...
+          repo.pulls.forEach((pull) => {
+            // I will decide which properties to use, such as user.login, name, and title, by looking at the actual Dependabot data.
+            if (pull.user.login === DEPENDABOT_USER_NAME) {
+              console.log(`c.log ## pull ##`, pull)
+              selectAllDependabot(repo.owner, repo.name, pull.number)
+            }
+          })
+        }
+      })
+    }
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:57]: let's do if (!repos) { return } to reduce nesting.

File: src/modules/github/contexts/selected-pulls-provider.tsx [hunk 8]
```diff
@@ -327,6 +335,51 @@ const selectedPullsReducer = (
...
+                return pull
+              })
+            }
+          }
+
+          return repo
+        })
+      }
+    }
+
```
Comment (jiyeol-lee) [inline src/modules/github/contexts/selected-pulls-provider.tsx:382]: try to write test code about this

**Comment by jiyeol-lee** [comment]
@jisung-lee7 What about we disable the menu if there is no dependabot pull request?

**Comment by jisung-lee7** [comment]
> @jisung-lee7 How about we disable the menu if there is no dependabot pull request?

@jiyeol-lee Of course, I can~!

File: src/modules/github/contexts/selected-pulls-provider.test.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
+import {
+  type TSelectedPullActionType,
+  type TState,
```
Comment (jiyeol-lee) [inline src/modules/github/contexts/selected-pulls-provider.test.tsx:3]: Nope, don't need to care about type when you are testing

File: src/modules/github/contexts/selected-pulls-provider.tsx [hunk 10]
```diff
@@ -14,11 +14,11 @@ import type {
   RepoPullsResponse
 } from '@/types/github/root/server'
 
-type TState = {
+export type TState = {
   repos: TRepo[] | undefined
 }
 
-type TSelectedPullActionType =
+export type TSelectedPullActionType =
```
Comment (jiyeol-lee) [inline src/modules/github/contexts/selected-pulls-provider.tsx:21]: Remove `export` if this is for test code

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 11]
```diff
@@ -29,25 +41,96 @@ export default function Page() {
...
+
+    const hasDependabot = repos.some((repo) => {
+      if (!repo.pulls) {
+        return
+      }
+      return repo.pulls.some((pull) => pull.user.login === DEPENDABOT_USER_NAME)
+    })
+
+    setHasDependabotPulls(hasDependabot)
+  }, [repos])
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:111]: You do not need another state for this

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 12]
```diff
@@ -14,12 +17,21 @@ import {
...
+
+const DEPENDABOT_USER_NAME = 'dependabot[bot]'
+
 export default function Page() {
   const { repos } = useRepos()
   const flattenCheckedPulls = getFlattenCheckedPulls(repos)
   const { translate } = useDictionary()
   const [isDialogOpen, setIsDialogOpen] = useState<boolean>(false)
-  const { openAllRepo } = useUpdateRepoOrPull()
+  const [hasDependabotPulls, setHasDependabotPulls] = useState<boolean>(false)
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:33]: No need `const [hasDependabotPulls, setHasDependabotPulls] = useState<boolean>(false)`. If you are having more state that is not needed, it's going to be difficult to debug in the future

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 13]
```diff
@@ -29,25 +41,96 @@ export default function Page() {
...
+  }
+
+  useEffect(() => {
+    if (!repos) {
+      return
+    }
+
+    const hasDependabot = repos.some((repo) => {
+      if (!repo.pulls) {
+        return
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:105]: return "what"? If you are doing like this, it will return `undefined`

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:105]: @jiyeol-lee  Ah....I'm 병신..;

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx [hunk 14]
```diff
@@ -29,25 +40,88 @@ export default function Page() {
...
+      label: translate('GITHUB.SELECT_ALL_DEPENDABOT_PULL_REQUESTS'),
+      value: ESettingsCode.SELECT_ALL_DEPENDABOT_PULL_REQUESTS,
+      disabled:
+        repos === undefined ||
+        repos.every(
+          (repo) =>
+            !repo.pulls?.some(
+              (pull) => pull.user.login === DEPENDABOT_USER_NAME
+            )
+        )
```
Comment (jiyeol-lee) [inline src/app/(views)/github/t/@pullButtons/pulls/page.tsx:75]:
```typescript
(repos ?? []).some((repo) => (repo.pulls ?? []).some((pull) => pull.user.login === DEPENDABOT_USER_NAME))
```

File: src/modules/github/contexts/selected-pulls-provider.tsx [hunk 15]
```diff
@@ -327,6 +335,51 @@ const selectedPullsReducer = (
...
+        return state
+      }
+
+      return {
+        ...state,
+        repos: state.repos.map((repo) => {
+          if (isTargetRepo(repo)) {
+            const isTargetPull = (pull: TPull) =>
+              pull.number === payload.pullNumber
+            const hasTargetPull = repo.pulls!.some((pull) => isTargetPull(pull))
```
Comment (jiyeol-lee) [inline src/modules/github/contexts/selected-pulls-provider.tsx:357]: How can you know `repo.pulls` are valid? Shouldn't you add exception logic for that?

File: src/modules/github/contexts/selected-pulls-provider.tsx [hunk 16]
```diff
@@ -327,6 +335,51 @@ const selectedPullsReducer = (
...
+
+            if (!hasTargetPull) {
+              return repo
+            }
+
+            return {
+              ...repo,
+              pulls: repo.pulls!.map((pull) => {
```
Comment (jiyeol-lee) [inline src/modules/github/contexts/selected-pulls-provider.tsx:365]: How can you know `repo.pulls` are valid? Shouldn't you add exception logic for that?

File: src/app/(views)/github/t/@pullButtons/pulls/page.tsx (hunk 14 above)
Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:75]: 이런 Nullish한 방법이?

Comment (jisung-lee7) [reply src/app/(views)/github/t/@pullButtons/pulls/page.tsx:75]: @jiyeol-lee 근데 Nullish는 알겠는데 만약 이렇게하면 한개라도 DEPENDABOT_USER_NAME이라면 disabled가 될텐데,,잘못된거아냐?

**Review by jiyeol-lee — APPROVED** [review]
It works as expected. Good job

**Comment by jisung-lee7** [comment]
#1

**Comment by jisung-lee7** [comment]
#113

**Comment by jisung-lee7** [comment]
@jisung-lee7 mentioned this pull request #364

**Pull request merged**